blips/blip-0002.mediawiki

```diff
@@ -0,0 +1,102 @@
+<pre>
+bLIP: 2
+Title: Keysend
+Status: Active
+Type: Standards Track
+Author: Valentine Wallace <vwallace@protonmail.com>
+Created: 2021-05-21
+License: BSD-2-Clause, OPL
+</pre>
+
+==Abstract==
+
+Keysend is a type of lightning payment that does not require the payee to
+provide an invoice. Instead, the payer includes their payer-selected payment
+preimage in the TLV onion.
+
+As of authoring this bLIP, keysend is already supported by lnd, C-Lightning,
+Eclair, LDK, and possibly others. This bLIP serves to document what is already
+well-supported in the wild, for posterity and so that new implementations don't
+have to reverse-engineer keysend from existing implementations.
+
+==Copyright==
+
+This bLIP is dual-licensed under the Open Publication License and BSD 2-clause
+license.
+
+==Specification==
+
+Sender:
+* MUST include a TLV record keyed by type `5482373484` with a TLV value of a
+  randomly generated and cryptographically-secure 32-byte value that serves as
+  the HTLC payment preimage
+* MUST NOT set a payment secret
+* SHOULD only send payments to nodes advertising feature bit 55 in the node
+  context starting June 1, 2022
+
+Receiver:
+* if completing the payment, MUST fulfill the HTLC using the TLV-provided
+  payment preimage
+* if failing the payment, SHOULD error with `PERM|15
```

Comment:
Should specify reason -> value, not just "if you fail for any reason, use this code".